catalog/recursion-to-filter.md

```diff
@@ -0,0 +1,87 @@
+# Recursion to Filter
+
+## Relevant To
+
+
+* Replace Loop with Collection Closure Method
+* Replace Temp with Chain
+
+## Relates To
+
+* Recursion to Map
+* Recursion to Reduce
+* Tail Call Optimise Recursion
+
+## Languages
+
+All
+
+## Motivation
+
+Selectively filter a list of items into a new, smaller list of items, based
+on defined criteria, in the form of a boolean statement or function.
+
+## Description
+
+The filter takes a function (called a predicate) that is passed each element from
+the original list. If the function returns `true` then that item is included in
+the new list. If it returns false then that item is not included.
+This is an immutable operation, returning a new list and not affecting the original.
+
+## Examples
+
+### Haskell
+
+#### Given
+
+```haskell
+numbers = [1, 6, 3, 5, 8, 67, 43, 4, 6, 8];
+```
+
+#### Before
+
+```haskell
+-- tail recursion
+filter' :: (a -> Bool) -> [a] -> [a]
+filter' _ [] = []
+filter' f (x:xs) = if (f x)
+                   then x : filter' f xs
+                   else     filter' f xs
+
+evenNumbers = filter' even numbers
+```
+
+#### After
+
+```haskell
+
+evenNumbers = filter even numbers
+```
+
+### PHP
+
+#### Given
+
+```php
+$numbers = [1, 6, 3, 5, 8, 67, 43, 4, 6, 8];
+```
+
+#### Before
+
+```php
+$evenNumbers = [];
+
+foreach ($numbers as $n) {
```

Comment:
I'm in 2 minds about using a `foreach` in this example. On one hand it's a very valid refactoring to recognise and apply when writing in PHP. In this case, maybe the refactoring name should be changed to *Recursion/Loop to Filter*. On the other hand, this is being presented as a catalogue of *Functional Programming Refactorings* and imperative loops are not something you'd expect to see in that context. What are your thoughts? Maybe there's an opportunity to show both and imperative and recursive example?